crates/bifrost/src/providers/replicated_loglet/replication/checker.rs

```diff
@@ -477,4 +1092,507 @@ mod tests {
             eq(FMajorityResult::None)
         );
     }
+
+    // Failure-domain-aware tests
+    #[test]
+    fn nodeset_checker_fd_basics() -> Result<()> {
+        // 3 nodes in each zone, and 3 zones in total in two regions
+        // additionally, we have 1 node without any region assignment
+        // total = 10 nodes
+        // region1
+        //      .az1  [N1, N2, N3]
+        //      .az2  [N4, N5, N6]
+        // region2
+        //      .az1  [N7, N8, N9]
+        // -         [N10] N11 - in Provisioning
+        let mut nodes_config = NodesConfiguration::new(Version::MIN, "test-cluster".to_owned());
+        // all authoritative
+        // region1.az1
+        for i in 1..=3 {
+            nodes_config.upsert_node(generate_logserver_node(
+                i,
+                StorageState::ReadWrite,
+                "region1.az1",
+            ));
+        }
+        // region1.az2
+        for i in 4..=6 {
+            nodes_config.upsert_node(generate_logserver_node(
+                i,
+                StorageState::ReadWrite,
+                "region1.az2",
+            ));
+        }
+        // region2.az1
+        for i in 7..=9 {
+            nodes_config.upsert_node(generate_logserver_node(
+                i,
+                StorageState::ReadWrite,
+                "region2.az1",
+            ));
+        }
+        // N10
+        nodes_config.upsert_node(generate_logserver_node(10, StorageState::ReadWrite, ""));
+        // N11
+        nodes_config.upsert_node(generate_logserver_node(11, StorageState::Provisioning, ""));
+
+        // # Scenario 1
+        // - Nodeset with all nodes = 10 nodes
+        // - Replication is {node: 3}
+        let nodeset: NodeSet = (1..=11).collect();
+        let replication = ReplicationProperty::new(3.try_into().unwrap());
+        let mut checker: NodeSetChecker<bool> =
+            NodeSetChecker::new(&nodeset, &nodes_config, &replication);
+        // all nodes in the nodeset are authoritative
+        assert_that!(checker.count_nodes(StorageState::is_authoritative), eq(10));
+        // all nodes are false by default. Can't establish write quorum.
+        assert_that!(checker.check_write_quorum(|attr| *attr), eq(false));
+        checker.set_attribute_on_each([1, 2, 10], true);
+        checker.set_attribute(10, false);
+        // trying 3 instead
+        checker.set_attribute(3, true);
```

Comment:
Could we have set [1, 2, 3] to true right away instead of going through 10?

Reply:
Good catch. The intention was to check with node 11 (not 10) and then add 3. I'll update.